### PR TITLE
MBP Touchbar Support

### PR DESCRIPTION
**Pull request by wiomoc**
Proof of Concept
This needs some work

### CODE REVIEW

**Comment by wiomoc** [comment]
touch bar bild 2017-08-09 um 20 21 22

**Comment by jpetrie** [comment]
Apart from some minor style nitpicks (I'll go through and leave comments on the specific lines), this mostly looks reasonable. I don't have the hardware to test it. When you say "this needs some work," what sort of work are you thinking of?

File: src/MacVim/MMVimController.m [hunk 1]
```diff
@@ -1306,7 +1355,53 @@ - (void)addToolbarItemWithLabel:(NSString *)label
...
+        // The label begins and ends with '-'; decided which kind of separator
+        // item it is by looking at the prefix.
+        if ([label hasPrefix:@"-space"]) {
+            label = NSTouchBarItemIdentifierFixedSpaceSmall;
+        } else if ([label hasPrefix:@"-flexspace"]) {
+            label = NSTouchBarItemIdentifierFlexibleSpace;
+        } else {
+            label = NSTouchBarItemIdentifierFixedSpaceLarge;
+        }
+    } else {
```
Comment (jpetrie) [inline src/MacVim/MMVimController.m]: This whole `else` block has mixed tab/space indenting and seemingly overly deep indents.

File: src/MacVim/MMWindowController.m [hunk 2]
```diff
@@ -1332,6 +1342,13 @@ - (void)runAfterWindowPresentedUsingBlock:(void (^)(void))block
     [afterWindowPresentedQueue addObject:[block copy]];
 }
 
+#if MAC_OS_X_VERSION_MAX_ALLOWED > MAC_OS_X_VERSION_10_12
+- (NSTouchBar *)makeTouchBar
+{
+	return [vimController makeTouchBar];
```
Comment (jpetrie) [inline src/MacVim/MMWindowController.m]: A tab is used to indent here instead of spaces.

**Comment by taiyangc** [comment]
Awesome idea! Is it possible to further user-customize the touch bar shortcuts to specific macros?

**Comment by getaaron** [comment]
@wiomoc @jpetrie any updates on this? would love to see it merged

**Comment by ychin** [comment]
I have the same question on updates. I'm quite interested in having TouchBar support as right now it doesn't do anything when I have MacVim up :(. (macOS allows you to force it to only show function keys when MacVim is up but if you do that it's hard to adjust volume, etc)

I don't mind taking this over and help push this through if the original author is no longer interested. I think this at least needs some documentation in `gui_mac.txt` in addition to the existing comments.

Played around with it a little bit, and some additional potential improvements below. These are really just random ideas. Could be future pull requests after this is in:

1. You can't pan through the touch bar items if there are too many. Ideally you should be able to do that.

* An alternative is expose the "Customize TouchBar…" option to give native TouchBar customization. Not sure how much work that is.

2. Different modes are not respected right now. E.g. `vnoremenu TouchBar.VisualOnly <Nop>` would still be active in normal mode. Normal menus and toolbars do respect the modes and gray out the items. I think for touch bar we don't even need to gray out the menu items. Just hide them to save space.
3. Nesting is not allowed right now. I.e. you cannot do `:an TouchBar.Grouping.Item1 <Nop>`. I'm thinking it could surface "Grouping" and when you tap on it it will show the individual items under it. This is useful as TouchBar's real estate is pretty small. E.g. you can use this to group debugger controls.

* Another idea is to show all the subitems, but grouped together. This saves the extra tap and allow you to see what the options are.

4. Color support. Touchbar items can have customized colors rather than the default ones.

5. Kind of extra credit, but support other GUI widgets. Maybe this gets a little too outside of what Vim usually is like, but examples range from using the standard template images (https://developer.apple.com/design/human-interface-guidelines/macos/touch-bar/touch-bar-icons-and-images/) to supporting scrubber. You can kind of go down a rabbit hole on this one though when trying to shoehorn support for native TouchBar APIs onto Vim APIs.

**Comment by jpetrie** [comment]
@ychin 

> I don't mind taking this over and help push this through if the original author is no longer interested.

Given the amount of time this PR has sat here I think it would be fine for you to do so. As I noted in my initial review, most of what I'd want to see fixed before merging are just style issues to try and keep the code consistent. If you can build a PR that incorporates this original change and fixes those, it would probably get merged pretty quickly.

I don't have hardware to test it on, and I'm not _personally_ that interested in the feature, but I _would_ like to see more contributors since I don't have much time to devote myself. So by all means, if you want to pick this up and run with it, please do so!

**Comment by ychin** [comment]
I just submitted a new PR (#715) which takes this current patch and fixed up misc issues and added features.

**Pull request closed without merging**